index.d.ts

```diff
@@ -26,28 +23,37 @@ declare type UnionToIntersection<Union> = (
 declare type AnyFunction = (...args: any) => any;
 declare type ReturnTypeOf<T extends AnyFunction | AnyFunction[]> =
   T extends AnyFunction
-    ? ReturnType<T>
-    : T extends AnyFunction[]
-    ? UnionToIntersection<Exclude<ReturnType<T[number]>, void>>
-    : never;
+  ? ReturnType<T>
+  : T extends AnyFunction[]
+  ? UnionToIntersection<Exclude<ReturnType<T[number]>, void>>
+  : never;
 
 type ClassWithPlugins = Constructor<any> & {
-  plugins: any[];
+  plugins: Plugin[];
 };
 
+type RemainingRequirements<PredefinedOptions> =
+  keyof PredefinedOptions extends never
+  ? Base.Options
+  : Omit<Base.Options, keyof PredefinedOptions>
+
+type NonOptionalKeys<Obj> = {
+  [K in keyof Obj]: {} extends Pick<Obj, K> ? undefined : K;
+}[keyof Obj];
+
+type RequiredIfRemaining<PredefinedOptions, NowProvided> =
+  NonOptionalKeys<RemainingRequirements<PredefinedOptions>> extends undefined
```

Comment:
This was an odd little issue. `tsd` was giving different issues the [different errors](https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions/pull/63#issuecomment-882881453) than VS Code's intellisense. My best guess for a root cause would be something about different compiler options in `tsd'` defaults vs. VS Code's when there isn't a `tsconfig.json`.

I debugged by adding a `debugKeys: NonOptionalKeys<RemainingRequirements<PredefinedOptions>>;`  in the class returned by `ConstructorRequiringVersion`'s `new`: https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions/pull/63/commits/32744acc553115fbde6c6ee1de651842b36e86bb

...which showed itself to be type `never` in VS Code but type `undefined` in `tsd`. 💡 

I switched from `never` to `undefined` to make sure both variants of possible compiler options work.

Reply:
> My best guess for a root cause would be something about different compiler options in `tsd'` defaults

just fyi, here is `tsd`'s default config:

> ```json
> {
> 	"strict": true,
> 	"jsx": "react",
> 	"target": "es2017",
> 	"lib": ["es2017"],
> 	"module": "commonjs",
> 	// The following option is set and is not overridable:
> 	"moduleResolution": "node"
> }
> ```

https://github.com/SamVerschueren/tsd#custom-typescript-config